File: courses/developingapps/datastore/start/server/gcp/datastore.js
// Copyright 2017, Google, Inc.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';



// Lists all questions in a Quiz (defaults to 'gcp').
// Returns a promise
// [START list]
function list(quiz = 'gcp') {
  // Placeholder statement
  return Promise.resolve({
    questions: [{
      id: -1,
      title: 'Dummy question',
      answer1: 'Dummy answer1',
      answer2: 'Dummy answer2',
      answer3: 'Dummy answer3',
      answer4: 'Dummy answer4',
      correctAnswer: 2
    }],
    nextPageToken: 'NO_MORE_RESULTS'
  });
}
// [END list]

// [START create]
function create({ quiz, author, title, answer1, answer2, answer3, answer4, correctAnswer}) {
  // Placeholder statement
  return Promise.resolve({});
}
// [END create]

// [START exports]
module.exports = {
  create,
  list
};
// [END exports]

File: courses/developingapps/v1.3/python/datastore/start/quiz/gcp/datastore.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# TODO: Import the os module



# END TODO

# TODO: Get the GCLOUD_PROJECT environment variable



# END TODO

from flask import current_app

# TODO: Import the datastore module from the google.cloud package



# END TODO

# TODO: Create a Cloud Datastore client object
# The datastore client object requires the Project ID.
# Pass through the Project ID you looked up from the
# environment variable earlier



# END TODO

"""
Returns a list of question entities for a given quiz
- filter by quiz name, defaulting to gcp
- no paging
- add in the entity key as the id property 
- if redact is true, remove the correctAnswer property from each entity
"""
def list_entities(quiz='gcp', redact=True):
    return [{'quiz':'gcp', 'title':'Sample question', 'answer1': 'A', 'answer2': 'B', 'answer3': 'C', 'answer4': 'D', 'correctAnswer': 1, 'author': 'Nigel'}]

"""
Create and persist and entity for each question
The Datastore key is the equivalent of a primary key in a relational database.
There are two main ways of writing a key:
1. Specify the kind, and let Datastore generate a unique numeric id
2. Specify the kind and a unique string id
"""
def save_question(question):
# TODO: Create a key for a Datastore entity whose kind is Question
    pass
    

# END TODO

# TODO: Create a Datastore entity object using the key

    

# END TODO

# TODO: Iterate over the form values supplied to the function

    

# END TODO

# TODO: Assign each key and value to the Datastore entity

        

# END TODO


# TODO: Save the entity

    

# END TODO